### PR TITLE
fix can't save image by image id

### PR DESCRIPTION
**Pull request by ningmingxiao**
fix can't save image by image id

### CODE REVIEW

File: cmd/nerdctl/save.go [hunk 1]
```diff
@@ -17,106 +17,124 @@
...
+                Client: client,
+                OnFound: func(ctx context.Context, found imagewalker.Found) error {
+                        if found.Image.Name != "" {
+                                locaImageNames = append(locaImageNames, found.Image.Name)
+                        }
+                        return nil
+                },
+        }
+        for _, img := range images {
+                walker.Walk(ctx, img)
```
Comment (AkihiroSuda) [inline cmd/nerdctl/save.go:121]: catch err

File: cmd/nerdctl/save.go [hunk 2]
```diff
@@ -17,106 +17,124 @@
...
+                                return err
+                        }
+                        saveOpts = append(saveOpts, archive.WithImage(imageStore, named.String()))
+                } else {
+                        for _, localName := range locaImageNames {
+                                saveOpts = append(saveOpts, archive.WithImage(imageStore, localName))
+                        }
+                }
+        }
+        return client.Export(ctx, out, saveOpts...)
```
Comment (AkihiroSuda) [inline cmd/nerdctl/save.go:134]: you need to call `Export` for every matched image string

Comment (ningmingxiao) [reply cmd/nerdctl/save.go]: if there are two images have same id, i will export all of them

File: cmd/nerdctl/save.go [hunk 3]
```diff
@@ -105,17 +105,60 @@ func saveImage(images []string, out io.Writer, saveOpts []archive.ExportOpt, cmd
...
 	for _, img := range images {
-		named, err := referenceutil.ParseAny(img)
+		foundImageNames = empty
+		foundId = empty
+		_, err = walker.Walk(ctx, img)
 		if err != nil {
 			return err
 		}
-		saveOpts = append(saveOpts, archive.WithImage(imageStore, named.String()))
+		if err = checkRepeatId(foundId); err != nil {
```
Comment (AkihiroSuda) [inline cmd/nerdctl/save.go:130]: Just check the ret value of `walker.Walk`

File: cmd/nerdctl/save.go [hunk 4]
```diff
@@ -105,17 +105,62 @@ func saveImage(images []string, out io.Writer, saveOpts []archive.ExportOpt, cmd
...
+	var foundId []string
+	walker := &imagewalker.ImageWalker{
+		Client: client,
+		OnFound: func(ctx context.Context, found imagewalker.Found) error {
+			imgName := found.Image.Name
+			imgDigest := found.Image.Target.Digest.String()
+			if imgName != "" && imgDigest != "" {
+				foundImageNames = append(foundImageNames, imgName)
+				foundId = append(foundId, imgDigest)
+				if err = checkRepeatId(foundId); err != nil {
```
Comment (junnplus) [inline cmd/nerdctl/save.go:118]: Please check `found.MatchCount`

Comment (junnplus) [inline cmd/nerdctl/save.go:118]: Please check `found.MatchCount`

File: cmd/nerdctl/save.go [hunk 5]
```diff
@@ -105,17 +105,62 @@ func saveImage(images []string, out io.Writer, saveOpts []archive.ExportOpt, cmd
...
-		saveOpts = append(saveOpts, archive.WithImage(imageStore, named.String()))
+		for _, localName := range foundImageNames {
+			saveOpts = append(saveOpts, archive.WithImage(imageStore, localName))
+		}
 	}
-
 	return client.Export(ctx, out, saveOpts...)
 }
 
+func checkRepeatId(stringSlice []string) error {
```
[the remark this answers is not in the document]
Comment (ningmingxiao) [reply cmd/nerdctl/save.go:142]: for example 
```
nerdctl images 
127.0.0.1:5000/pause    3.5              2f4b437353f9    2 weeks ago           linux/amd64    676.0 KiB    291.0 KiB
busybox                 stable-musl      2a3b33f1604b    2 seconds ago         linux/amd64    1.4 MiB      827.5 KiB
```
use nerdctl save 2 -o test.tar will not show error， if i delete  checkRepeatId

[the remark this answers is not in the document]
Comment (ningmingxiao) [reply cmd/nerdctl/save.go:142]: i see  use found.MatchCount instead of checkRepeatId

File: cmd/nerdctl/save.go [hunk 6]
```diff
@@ -105,12 +105,36 @@ func saveImage(images []string, out io.Writer, saveOpts []archive.ExportOpt, cmd
...
+			} else {
+				return fmt.Errorf("imgName or imgDigest is null")
+			}
+			return nil
+		},
+	}
+	var empty []string
 	for _, img := range images {
-		named, err := referenceutil.ParseAny(img)
+		foundImageNames = empty
```
Comment (junnplus) [inline cmd/nerdctl/save.go:126]: why?

File: cmd/nerdctl/save.go [hunk 7]
```diff
@@ -105,12 +105,36 @@ func saveImage(images []string, out io.Writer, saveOpts []archive.ExportOpt, cmd
...
+		foundID = empty
+		count, err := walker.Walk(ctx, img)
 		if err != nil {
 			return err
 		}
-		saveOpts = append(saveOpts, archive.WithImage(imageStore, named.String()))
+		if count > 1 {
+			return fmt.Errorf("ambiguous digest id")
+		}
+		for _, localName := range foundImageNames {
```
Comment (junnplus) [inline cmd/nerdctl/save.go:135]: Move it outside the for loop of images.

**Comment by ningmingxiao** [comment]
if  use walker.Walk(ctx, img)  to check count will show error if an image have different tags @junnplus

File: cmd/nerdctl/save.go [hunk 8]
```diff
@@ -105,17 +105,59 @@ func saveImage(images []string, out io.Writer, saveOpts []archive.ExportOpt, cmd
...
+	var foundID []string
+	walker := &imagewalker.ImageWalker{
+		Client: client,
+		OnFound: func(ctx context.Context, found imagewalker.Found) error {
+			imgName := found.Image.Name
+			imgDigest := found.Image.Target.Digest.String()
+			if imgName != "" && imgDigest != "" {
+				foundImageNames = append(foundImageNames, imgName)
+				foundID = append(foundID, imgDigest)
+				if err = checkRepeatID(foundID); err != nil {
```
Comment (junnplus) [inline cmd/nerdctl/save.go:118]: Why need it?

File: cmd/nerdctl/save.go [hunk 9]
```diff
@@ -105,17 +105,59 @@ func saveImage(images []string, out io.Writer, saveOpts []archive.ExportOpt, cmd
...
+		_, err := walker.Walk(ctx, img)
 		if err != nil {
 			return err
 		}
-		saveOpts = append(saveOpts, archive.WithImage(imageStore, named.String()))
 	}
-
+	for _, localName := range foundImageNames {
+		saveOpts = append(saveOpts, archive.WithImage(imageStore, localName))
+	}
```
Comment (junnplus) [inline cmd/nerdctl/save.go:135]: Put it into `OnFound` of the walker.

File: cmd/nerdctl/save.go [hunk 10]
```diff
@@ -105,17 +105,62 @@ func saveImage(images []string, out io.Writer, saveOpts []archive.ExportOpt, cmd
...
 	for _, img := range images {
-		named, err := referenceutil.ParseAny(img)
+		count, err := walker.Walk(ctx, img)
 		if err != nil {
 			return err
 		}
-		saveOpts = append(saveOpts, archive.WithImage(imageStore, named.String()))
+		if count == 0 {
+			return fmt.Errorf("no such image:%s", img)
```
Comment (junnplus) [inline cmd/nerdctl/save.go:133]:
```suggestion
			return fmt.Errorf("no such image: %s", img)
```

File: cmd/nerdctl/save.go (hunk 8 above)
Comment (ningmingxiao) [reply cmd/nerdctl/save.go:118]:
```
[root@k8s nmx]# nerdctl  -n k8s.io images|grep bc
calico/node                           v3.19.1                                                             bc4a631d553b    28 hours ago    linux/amd64    160.6 MiB    56.0 MiB
calico/pod2daemon-flexvol             v3.19.1                                                             bcac7dc4f130    28 hours ago    linux/amd64    20.7 MiB     8.9 MiB
sha256                                5660150975fb82f006d013fdb583b628d3d340ff64775ca3111ca28f523e32df    bcac7dc4f130    28 hours ago    linux/amd64    20.7 MiB     8.9 MiB
sha256                                c4d75af7e098eef5142c65b8fab5b3ff0e5c614ce72e91b8c44ba20249b596ef    bc4a631d553b    28 hours ago    linux/amd64    160.6 MiB    56.0 MiB
```
if i delete it 
nerdctl  -n k8s.io save bca -o test.tar will not show error

Comment (junnplus) [reply cmd/nerdctl/save.go:119]:
```go
 108     var foundID map[string]struct{}
 109     walker := &imagewalker.ImageWalker{
 110         Client: client,
 111         OnFound: func(ctx context.Context, found imagewalker.Found) error {
 112             imgName := found.Image.Name
 113             imgDigest := found.Image.Target.Digest.String()
 114             if _, ok := foundID[imgDigest]; !ok && len(foundID) == 1 {
 115                 return fmt.Errorf("ambiguous digest string")
 116             }
 117
 118             foundID[imgDigest] = struct{}{}
 119             saveOpts = append(saveOpts, archive.WithImage(imageStore, imgName))
 120             return nil
 121         },
 122     }
 123     for _, img := range images {
 124         foundID = make(map[string]struct{})
 125         n, err := walker.Walk(ctx, img)
 126         if err != nil {
 127             return err
 128         }
 129         if n == 0 {
 130             return fmt.Errorf("no such image: %s", img)
 131         }
 132     }
```

Comment (ningmingxiao) [reply cmd/nerdctl/save.go:119]:
```
[root@localhost _output]# ./nerdctl images
REPOSITORY                   TAG                                                                 IMAGE ID        CREATED              PLATFORM       SIZE         BLOB SIZE
calico/kube-controllers      v3.19.1                                                             904458fe1bd5    About an hour ago    linux/amd64    57.8 MiB     23.9 MiB
calico/node                  v3.19.1                                                             bc4a631d553b    24 minutes ago       linux/amd64    165.3 MiB    56.0 MiB
calico/pod2daemon-flexvol    v3.19.1                                                             bcac7dc4f130    24 minutes ago       linux/amd64    20.8 MiB     8.9 MiB
k8s.gcr.io/pause             3.5                                                                 1ff6c18fbef2    About an hour ago    linux/amd64    668.0 KiB    294.4 KiB
sha256                       5660150975fb82f006d013fdb583b628d3d340ff64775ca3111ca28f523e32df    bcac7dc4f130    24 minutes ago       linux/amd64    20.8 MiB     8.9 MiB
sha256                       5d3d5ddc8605ded8f69d76ee488072c7d02c32a8e4e8b34640a884c6eb939c0a    904458fe1bd5    About an hour ago    linux/amd64    57.8 MiB     23.9 MiB
sha256                       c4d75af7e098eef5142c65b8fab5b3ff0e5c614ce72e91b8c44ba20249b596ef    bc4a631d553b    24 minutes ago       linux/amd64    165.3 MiB    56.0 MiB
sha256                       ed210e3e4a5bae1237f1bb44d72a05a2f1e5c6bfe7a7e73da179e2534269c459    1ff6c18fbef2    About an hour ago    linux/amd64    668.0 KiB    294.4 KiB

```
if use nerdctl save bc 1f -o test.tar will not work (four ID start with bc  )

Comment (junnplus) [reply cmd/nerdctl/save.go:119]: Keep same with docker.

Comment (fahedouch) [reply cmd/nerdctl/save.go]: `nerdctl load` support multi image import so we should support multi-image save

File: cmd/nerdctl/save.go [hunk 11]
```diff
@@ -103,19 +102,47 @@ func saveImage(images []string, out io.Writer, saveOpts []archive.ExportOpt, cmd
...
+		count, err := walker.Walk(ctx, img)
 		if err != nil {
 			return err
 		}
-		saveOpts = append(saveOpts, archive.WithImage(imageStore, named.String()))
+		if count == 0 {
+			return fmt.Errorf("no such image: %s", img)
+		}
+	}
+	err = checkAmbiguousID(images, foundID)
```
Comment (fahedouch) [inline cmd/nerdctl/save.go:129]: `nerdctl load` support multi-image load so no need for this

File: cmd/nerdctl/save.go [hunk 12]
```diff
@@ -103,16 +103,36 @@ func saveImage(images []string, out io.Writer, saveOpts []archive.ExportOpt, cmd
...
+		if count == 0 {
+			return fmt.Errorf("no such image: %s", img)
+		}
+	}
+	if len(foundID) > len(images) {
+		return fmt.Errorf("ambiguous digest id")
+	}
+	if err != nil {
+		return err
 	}
```
Comment (junnplus) [inline cmd/nerdctl/save.go:135]: ?

File: cmd/nerdctl/save.go [hunk 13]
```diff
@@ -83,6 +83,7 @@ func saveAction(cmd *cobra.Command, args []string) error {
 }
 
 func saveImage(images []string, out io.Writer, saveOpts []archive.ExportOpt, cmd *cobra.Command) error {
+	images = uniqueSlice(images)
```
Comment (junnplus) [inline cmd/nerdctl/save.go:86]: strutil.DedupeStrSlice

File: cmd/nerdctl/save_linux_test.go [hunk 14]
```diff
@@ -48,6 +48,25 @@ func TestSave(t *testing.T) {
 	assert.Assert(t, strings.Contains(etcOSRelease, "Alpine"))
 }
 
+func TestSaveById(t *testing.T) {
+	base := testutil.NewBase(t)
+	base.Cmd("pull", testutil.GolangImage).AssertOK()
```
Comment (fahedouch) [inline cmd/nerdctl/save_linux_test.go:53]: plz use  `testutil.AlpineImage` , I believe it is smaller

**Comment by ningmingxiao** [comment]
thank you @junnplus

File: cmd/nerdctl/save_linux_test.go [hunk 15]
```diff
@@ -48,6 +48,25 @@ func TestSave(t *testing.T) {
...
+	base.Cmd("pull", testutil.AlpineImage).AssertOK()
+	var format string
+	if testutil.GetTarget() == testutil.Docker {
+		format = "{{.Id}}"
+	} else {
+		format = "{{.RepoDigests}}"
+	}
+	res := base.Cmd("inspect", testutil.AlpineImage, "--format", format).Run()
+	repoDigests := strings.Split(res.Stdout(), ":")[1]
+	id := repoDigests[0 : len(repoDigests)-2]
```
Comment (junnplus) [inline cmd/nerdctl/save_linux_test.go:62]: using `base.InspectImage`

**Review by junnplus — APPROVED** [review]
Thanks, LGTM on green

File: cmd/nerdctl/save.go [hunk 16]
```diff
@@ -103,16 +105,33 @@ func saveImage(images []string, out io.Writer, saveOpts []archive.ExportOpt, cmd
...
 		if err != nil {
 			return err
 		}
-		saveOpts = append(saveOpts, archive.WithImage(imageStore, named.String()))
+		if count == 0 {
+			return fmt.Errorf("no such image: %s", img)
+		}
+	}
+	if len(foundID) > len(images) {
+		return fmt.Errorf("ambiguous digest id")
```
Comment (fahedouch) [inline cmd/nerdctl/save.go:133]: not sure but I think this is useless

Comment (ningmingxiao) [reply cmd/nerdctl/save.go:133]:
```
[root@k8s nmx]# nerdctl  -n k8s.io images|grep bc
calico/node                           v3.19.1                                                             bc4a631d553b    28 hours ago    linux/amd64    160.6 MiB    56.0 MiB
calico/pod2daemon-flexvol             v3.19.1                                                             bcac7dc4f130    28 hours ago    linux/amd64    20.7 MiB     8.9 MiB
sha256                                5660150975fb82f006d013fdb583b628d3d340ff64775ca3111ca28f523e32df    bcac7dc4f130    28 hours ago    linux/amd64    20.7 MiB     8.9 MiB
sha256                                c4d75af7e098eef5142c65b8fab5b3ff0e5c614ce72e91b8c44ba20249b596ef    bc4a631d553b    28 hours ago    linux/amd64    160.6 MiB    56.0 MiB
```
for example nerdctl -n k8s.io save  bc   -o test.tar  ,I want to show error.if not use this it wll not show error

Comment (AkihiroSuda) [reply cmd/nerdctl/save.go:133]: Checking `found.MatchCount > 1` might be enough

Comment (AkihiroSuda) [reply cmd/nerdctl/save.go:133]: ping

**Comment by djdongjin** [comment]
Probably I can give a hand so the PR can be merged :) cc @AkihiroSuda

**Comment by djdongjin** [comment]
close as #1776 is merged. thanks

**Pull request closed without merging**